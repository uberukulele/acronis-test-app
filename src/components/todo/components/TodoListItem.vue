<template>
  <div :class="b()">
    <el-checkbox :value="todo.done"
                 @input="toggleTodoDone"/>
    <span :class="b('text')"
          :title="todo.text">{{ todo.text }}</span>
    <el-tooltip :content="$t('todos.delete-tooltip')"
                effect="dark"
                placement="bottom-start">
      <i :class="b('remove-icon','el-icon-close')"
         @click="maybeDeleteTodo"/>
    </el-tooltip>

  </div>
</template>

<script>
  import { mapActions } from 'vuex'
  import { Checkbox, Tooltip } from 'element-ui'

  export default {
    name: 'todo-list-item',
    props: {
      todo: {
        type: Object,
        required: true
      }
    },
    components: {
      ElCheckbox: Checkbox,
      ElTooltip: Tooltip
    },
    methods: {
      ...mapActions(['updateTodo', 'deleteTodo']),
      toggleTodoDone (done) {
        this.updateTodo({
          ...this.todo,
          done
        })
      },

      async maybeDeleteTodo () {
        try {
          await this.$confirm(this.$t('todos.delete-todo-q'),
              this.$t('common.confirm-title'), {
                confirmButtonText: this.$t('common.buttons.delete'),
                cancelButtonText: this.$t('common.buttons.cancel'),
                type: 'warning'
              })
          await this.deleteTodo(this.todo.id)
        } catch (e) {
          console.log(e)
        }
      }
    }
  }
</script>

<style lang="stylus">
  @import "~@/theme/variables.styl"

  .todo-list-item {
    display flex
    align-items center
    height 52px
    padding 0 20px
    border-bottom 1px solid $borderColor

    &__text {
      flex 1
      padding-left 20px
      overflow hidden
      white-space nowrap
      text-overflow ellipsis
      font-size 14px
      color $textColor
    }

    &__remove-icon {
      display none
      padding 6px
      cursor pointer
      margin-right -6px
      font-size 18px
      color $primary
    }

    &:hover {
      background-color rgba($primary, 0.1)
    }

    &:hover &__remove-icon {
      display block
    }
  }
</style>
