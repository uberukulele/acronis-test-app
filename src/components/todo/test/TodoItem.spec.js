import { mount, shallowMount } from '@vue/test-utils'
import TodoListItem from '@/components/todo/components/TodoListItem.vue'

const mockedUpdateTodo = jest.spyOn(TodoListItem.methods, 'updateTodo')
const mockedDeleteTodo = jest.spyOn(TodoListItem.methods, 'maybeDeleteTodo')

const getWrapper = (shallow = false) => {
  let todos = [{
    id: 1,
    text: 'Test',
    done: false
  }, {
    id: 2,
    text: 'Test 1',
    done: true
  }]

  let mountMethod = shallow ? shallowMount : mount

  return mountMethod(TodoListItem, {
    propsData: {
      todo: todos[1]
    },
    mocks: {
      $t: msg => msg
    }
  })
}

describe('TodoListItem', () => {
  it('Should render', () => {
    expect(getWrapper().find('.todo-list-item').isVisible()).toBeTruthy()
  })

  it('Should call updateTodo on checkbox click', () => {
    let wrapper = getWrapper()
    wrapper.find('.el-checkbox input').trigger('click')
    expect(mockedUpdateTodo).toHaveBeenCalled()
  })

  it('Should call maybeDeleteTodo on delete btn click', () => {
    let wrapper = getWrapper()
    wrapper.find('.el-icon-close').trigger('click')
    expect(mockedDeleteTodo).toHaveBeenCalled()
  })

  it('Snapshot test', () => {
    const wrapper = getWrapper(true)
    expect(wrapper.element).toMatchSnapshot()
  })
})
