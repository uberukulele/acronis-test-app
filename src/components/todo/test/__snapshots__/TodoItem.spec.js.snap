// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoListItem renders correctly 1`] = `
<div
  class="todo-list-item"
>
  <el-checkbox-stub
    value="true"
  />
   
  <span
    class="todo-list-item__text"
    title="Test 1"
  >
    Test 1
  </span>
   
  <el-tooltip-stub
    appendtobody="true"
    arrowoffset="0"
    boundariespadding="5"
    content="todos.delete-tooltip"
    effect="dark"
    enterable="true"
    hideafter="0"
    offset="0"
    opendelay="0"
    placement="bottom-start"
    popperoptions="[object Object]"
    tabindex="0"
    transformorigin="true"
    transition="el-fade-in-linear"
    visiblearrow="true"
  >
    <i
      class="todo-list-item__remove-icon el-icon-close"
    />
  </el-tooltip-stub>
</div>
`;
